service.go

```diff
@@ -103,26 +125,21 @@ func (client *Client) runRequest(req *http.Request) ([]byte, error) {
 
 	RateLimiting.updateRateLimits(resp)
 
-	return checkResponseForErrors(resp)
+	return checkResponseForErrorsWithErroHandler(resp, errorHandler)
 }
 
-func checkResponseForErrors(resp *http.Response) ([]byte, error) {
-	// check status code, could be 500, or most likely the client_secret is incorrect
-	if resp.StatusCode/100 == 5 {
-		return nil, errors.New("server error")
-	}
-
-	if resp.StatusCode/100 == 4 {
-		var response Error
-		contents, _ := ioutil.ReadAll(resp.Body)
-		json.Unmarshal(contents, &response)
-
-		return nil, response
-	}
+func (client *Client) runRequest(req *http.Request) ([]byte, error) {
+	return client.runRequestWithErrorHandler(req, defaultErrorHandler)
+}
 
-	if resp.StatusCode/100 == 3 {
-		return nil, errors.New("redirect error")
+func checkResponseForErrorsWithErroHandler(resp *http.Response, errorHandler ErrorHandler) ([]byte, error) {
```

Comment:
Missing r on the second error.  Otherwise looks good.  

We have a known issue where an upload post can return errors as "Upload Objects" with error message or 400 and an api error message. We should make it so that anything wrong that's determined initially should return the error object. And anything that happens during processing will be be on the upload when one polls for status.

FYI @mathewbgordon this is basically API-344